File: leetcode/988-smallest-string-starting-from-leaf.py
# https://leetcode.com/problems/smallest-string-starting-from-leaf/

from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        
        def dfs(node, path = []):
            if not node.left and not node.right:
                return path + [node.val]
            
            curr_path = path + [node.val]

            if node.left and node.right:
                return self.get_min(dfs(node.left, curr_path), dfs(node.right, curr_path))
            
            return dfs((node.right if node.right else node.left), curr_path)

        
        inverted_result = dfs(root)

        result = []

        for i in reversed(range(len(inverted_result))):
            result.append(chr(97 + inverted_result[i]))

        return "".join(result)

    def get_min(self, arr1, arr2):
        i = len(arr1) - 1
        j = len(arr2) - 1
        while i >= 0 and j >= 0:
            if arr1[i] < arr2[j]:
                return arr1
            elif arr2[j] < arr1[i]:
                return arr2
            i -= 1
            j -= 1
        
        return arr2 if j == -1 else arr1

File: leetcode/988-smallest-string-starting-from-leaf.js
// https://leetcode.com/problems/smallest-string-starting-from-leaf/

// Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {string}
 */
var smallestFromLeaf = function(root) {
  const getMin = (arr1, arr2) => {
    let i = arr1.length - 1;
    let j = arr2.length - 1;

    while (i >= 0 && j >= 0) {
      if (arr1[i] < arr2[j]) return arr1;
      else if (arr1[i] > arr2[j]) return arr2;
      i -= 1;
      j -= 1;
    }

    return j === -1 ? arr2 : arr1;
  }
  
  const dfs = (node, path = []) => {
    if (!node.left && !node.right) return [...path, node.val];

    const currPath = [...path, node.val];
    if (node.left && node.right) return getMin(dfs(node.left, currPath), dfs(node.right, currPath));
    return dfs((node.left || node.right), currPath);
  }

  const result = dfs(root);

  const resultChar = [];

  for (let i = result.length - 1; i >= 0; i -= 1) {
    resultChar.push(String.fromCharCode(97 + result[i]))
  }
  
  return resultChar.join('');
};
